### PR TITLE
custom channel connector docs improvements

### PR DESCRIPTION
**Pull request by melindaloubser1**
**Proposed changes**:
- Restructure the custom connector docs page to clarify how it should be implemented and referred to
- closes https://github.com/RasaHQ/rasa/issues/6626

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [x] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by melindaloubser1** [comment]
@erohmensing The description of calling `on_new_message` is still very confusing (to me). Where is `on_new_message` handled? I can't find it defined anywhere in the codebase, it seems to just get called by all the `blueprint` methods.

**Comment by melindaloubser1** [comment]
@erohmensing have you had a chance to look at this yet?

File: docs/docs/connectors/custom-connectors.mdx [hunk 1]
```diff
@@ -5,70 +5,199 @@ title: Custom Connectors
...
+and implement at least a `blueprint` and `name` method. 
 
-The `name` method defines the url prefix. For example, if your component is
-named `myio`, the webhook you can use to attach the external service is
+
+## The `name` method
+
+The `name` method defines the url prefix for the connector's webhook. It also defines the channel name you should use 
+in any [channel specific response variations](../responses.mdx#channel-specific-response-variations) and the name you
+should pass to the `output_channel` query parameter on the [trigger intent endpoint](https://www.rasa.com/docs/rasa/pages/http-api#operation/triggerConversationIntent)
```
Comment (erohmensing) [inline docs/docs/connectors/custom-connectors.mdx:20]:
```suggestion
should pass to the `output_channel` query parameter on the [trigger intent endpoint](https://www.rasa.com/docs/rasa/pages/http-api#operation/triggerConversationIntent).
```

File: docs/docs/connectors/custom-connectors.mdx [hunk 2]
```diff
@@ -5,70 +5,199 @@ title: Custom Connectors
...
+```
+
+You would write a response variation specific to the `myio` channel  as:
+
+```yaml-rasa title="domain.yml"
+responses:
+  utter_greet:
+    - text: Hi! I'm the default greeting.
+    - text: Hi! I'm the  custom channel greeting
+      channel: myio
```
Comment (erohmensing) [inline docs/docs/connectors/custom-connectors.mdx:40]:
```suggestion
  utter_greet:
    - text: Hi! I'm the default greeting.
    - text: Hi! I'm the custom channel greeting
      channel: myio
```

File: docs/docs/connectors/custom-connectors.mdx [hunk 3]
```diff
@@ -5,70 +5,199 @@ title: Custom Connectors
...
-  "message": "Hi there!",
-  "metadata": {}  // optional, any extra info you want to add for processing in NLU or custom actions
-}
+## The `blueprint` method
+
+The `blueprint` method
+needs to create a [sanic blueprint](https://sanic.readthedocs.io/en/v20.3.0/sanic/blueprints.html#blueprints)
+that can be attached to a sanic server. 
+Your blueprint should have at least the two routes: `health` on the route `/`,
+and `receive` on the route `/webhook`.
```
Comment (erohmensing) [inline docs/docs/connectors/custom-connectors.mdx:54]: I'd add here that there's an example further down, as this isn't immediately clear and the example requires a bit of scrolling to get to

File: docs/docs/connectors/custom-connectors.mdx [hunk 4]
```diff
@@ -5,70 +5,199 @@ title: Custom Connectors
...
-
-In your implementation of the `receive` endpoint, you will need
-to call `on_new_message(UserMessage(text, output, sender_id))`.
-This will tell Rasa Core to handle this user message. The `output`
-is an output channel implementing the `OutputChannel` class. You can
-either implement the methods for your particular chat channel (e.g. there
-are methods to send text and images) or you can use the
-`CollectingOutputChannel` to collect the bot responses Core
-creates while the bot is processing your messages and return
+See more details on the `UserMessage` object [here](https://www.rasa.com/docs/rasa/reference/rasa/core/channels/channel#usermessage-objects)
```
Comment (erohmensing) [inline docs/docs/connectors/custom-connectors.mdx:69]:
```suggestion
See more details on the `UserMessage` object [here](https://www.rasa.com/docs/rasa/reference/rasa/core/channels/channel#usermessage-objects).
```

File: docs/docs/connectors/custom-connectors.mdx [hunk 5]
```diff
@@ -5,70 +5,199 @@ title: Custom Connectors
...
+## The `blueprint` method
+
+The `blueprint` method
+needs to create a [sanic blueprint](https://sanic.readthedocs.io/en/v20.3.0/sanic/blueprints.html#blueprints)
+that can be attached to a sanic server. 
+Your blueprint should have at least the two routes: `health` on the route `/`,
+and `receive` on the route `/webhook`.
+
+As part of your implementation of the `receive` endpoint, you will need to tell
+Rasa core to handle the user message. You do this by calling
```
Comment (erohmensing) [inline docs/docs/connectors/custom-connectors.mdx:57]:
```suggestion
Rasa Core to handle the user message. You do this by calling
```

File: docs/docs/connectors/custom-connectors.mdx [hunk 6]
```diff
@@ -5,70 +5,199 @@ title: Custom Connectors
...
+## Metadata on messages
+
+If you need to use extra information from your front end in your custom
+actions, you can pass this information using the `metadata` key of your user
+message. This information will accompany the user message through the Rasa
+server into the action server when applicable, where you can find it stored in
+the `tracker`. Message metadata will not directly affect NLU classification
+or action prediction. 
+
+The `InputChannel` class provides a default implementation of the method `get_metadata`.
```
Comment (erohmensing) [inline docs/docs/connectors/custom-connectors.mdx:152]: maybe mention that the default implementation just ignores it 😅

**Review by erohmensing — APPROVED** [review]
This is so helpful! I'd definitely add a piece on where the `on_new_message` gets sent to bc its v unclear, just let me find it

**Comment by erohmensing** [comment]
I'm pretty sure this is where it is sent: https://github.com/RasaHQ/rasa/blob/main/rasa/core/agent.py#L511

When we define the blueprint, we give it a handler, which is a callable method that can handle the incoming requests (so in this case `on_new_message` is the handler). 

Then when we register the channels when loading up the server, we tell it that the handler method is going to refer specifically to rasa.core.agent.handle_message here: https://github.com/RasaHQ/rasa/blob/main/rasa/core/channels/channel.py#L85

At least that's how i understand it. I think it'll be clearer in the docs if we just explain that calling `on_new_message` will send that to handle_message

File: docs/docs/connectors/custom-connectors.mdx (hunk 5 above)
Comment (melindaloubser1) [reply docs/docs/connectors/custom-connectors.mdx:57]: This should actually just be "Rasa" right?

File: docs/docs/connectors/custom-connectors.mdx [hunk 7]
```diff
@@ -5,70 +5,200 @@ title: Custom Connectors
...
-
-In your implementation of the `receive` endpoint, you will need
-to call `on_new_message(UserMessage(text, output, sender_id))`.
-This will tell Rasa Core to handle this user message. The `output`
-is an output channel implementing the `OutputChannel` class. You can
-either implement the methods for your particular chat channel (e.g. there
-are methods to send text and images) or you can use the
-`CollectingOutputChannel` to collect the bot responses Core
-creates while the bot is processing your messages and return
+Calling `on_new_message` will send the user message to the [`handle_message`](https://github.com/RasaHQ/rasa/blob/c922253fe890bb4903329d4ade764e0711d384ec/rasa/core/agent.py#L511_) method.
```
Comment (melindaloubser1) [inline docs/docs/connectors/custom-connectors.mdx:69]: @erohmensing how's this?

**Comment by wochinge** [comment]
@melindaloubser1 Please merge to `main` or `2.3.x` instead of `2.2.x`. I'll release `2.3.0` today.

**Comment by melindaloubser1** [comment]
Ah shoot forgot this was still open

**Pull request merged**